### ISSUE TITLE
BERT for Patents: unable to access hidden layers

### ISSUE THREAD

**Issue by benzitohhh** [google/patents-public-data#45]
Hi there,

on the BERT for Patents "saved model", how can we access the hidden layers?

The response signature only contains "encoder_layer" (the final layer). But we would like to access all 24 layers.

Do we need to use the "checkpoint"? If so, how can we load this?

Thanks in advance!

**Comment by benzitohhh** [comment]
Ah we managed to access the hidden layers by explicitly loading the checkpoint - thanks!

**Comment by amirbar** [comment]
@benzitohhh could you please provide an example for this?

**Comment by benzitohhh** [comment]
@amirbar 

Using bert-for-tf2 (https://github.com/kpe/bert-for-tf2), the checkpoint can be loaded in TensorFlow 2 (Keras).

And then `out_layer_ndxs` can be set to access whatever layers you need.

i.e. below we access the penultimate layer (-2).

```
import bert
from tensorflow import keras
import tensorflow as tf

MODEL_DIR = "./bert_model"
MAX_SEQ_LEN = 256

l_input_ids = keras.layers.Input(shape=(MAX_SEQ_LEN,), dtype='int32')

bert_params = bert.params_from_pretrained_ckpt(MODEL_DIR)
bert_params.out_layer_ndxs = [-2]

l_bert = bert.BertModelLayer.from_params(bert_params, name="bert")
output = l_bert(l_input_ids)
model = keras.Model(inputs=l_input_ids, outputs=output)
model.build(input_shape=(None, MAX_SEQ_LEN))

# A simple input
texts             = ["<CLS>", "The", "quick", "brown", "fox",  "<SEP>"]
input_ids      = [ 2,             1661,   3913,     2494,     4084,  3      ]
padding        = [0] * (MAX_SEQ_LEN - len(input_ids))
input_ids_padded = input_ids + padding

result  = model.predict([input_ids_padded])

```

**Comment by avivihadar** [comment]
Hi @benzitohhh, thanks for providing the example. However, this seems to be missing the actual line for loading the weights (should be called after model.build):
`bert.load_bert_weights(l_bert, model_ckpt)`

However, the cls token weights do not seem to be defined in the target model. Did you attempt to redefine them on TF2?

**Comment by benzitohhh** [comment]
@avivihadar Ah thanks so much. You're totally right, I forgot to actually load the weights. Oooops :)

Maybe I'm misunderstanding, but the CLS token is just the first element in each sequence. So the weights should be there.

We actually have mainly been using bert-as-a-service (although it uses tensorflow1):
https://github.com/hanxiao/bert-as-service

**Comment by benzitohhh** [comment]
@avivihadar Oh wait, by CLS token, you mean the logits? Yeah we're not using that, we're just interested in getting a patent embedding (vectorisation).
